File: cpp/src/global_planning/bfs.cpp
#include "nav_study/global_planning/bfs.hpp"

#include <chrono>
#include <deque>
#include <unordered_map>
#include <unordered_set>

#include "nav_study/global_planning/discrete_search.hpp"

namespace nav_study::global_planning {

core::PlanResult<Cell> BfsPlanner::plan(DiscreteSpace<Cell>& space, const Cell& start,
                                        const Cell& goal, TraceRecorder* recorder) {
  auto t0 = std::chrono::steady_clock::now();
  std::deque<Cell> queue{start};
  std::unordered_set<Cell> discovered{start};
  std::unordered_map<Cell, Cell> came_from;
  std::unordered_map<Cell, double> cost_to{{start, 0.0}};
  PlanStats stats;
  bool found = false;

  while (!queue.empty()) {
    Cell u = queue.front();
    queue.pop_front();
    if (recorder) recorder->node_expanded(u, cost_to[u]);
    ++stats.expanded_nodes;
    if (u == goal) {
      found = true;
      break;
    }
    for (const auto& [v, w] : space.neighbors(u)) {
      if (discovered.count(v)) continue;  // first discovery gives the shortest-in-moves parent
      discovered.insert(v);
      came_from[v] = u;
      cost_to[v] = cost_to[u] + w;
      if (recorder) recorder->edge_added(v, u, w);
      queue.push_back(v);
    }
  }

  core::PlanResult<Cell> result;
  if (found) {
    result.success = true;
    result.path = reconstruct_path(came_from, start, goal);
    result.cost = cost_to[goal];  // true edge-cost sum along the fewest-edge path
    if (recorder) recorder->path_found(result.path);
  }
  result.stats = stats;
  double rt = std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();
  emit_finished_discrete(recorder, result.success, result.cost, stats, rt);
  return result;
}

}  // namespace nav_study::global_planning

File: python/nav_study/global_planning/bfs.py
"""Breadth-first search over a DiscreteSpace (Cormen et al.; fewest-edge path).

Each node is enqueued once at first discovery and its parent fixed to the
discoverer, which is what yields the fewest-edge path.
"""

from __future__ import annotations

import time
from collections import deque

from nav_study.core.capabilities import Capability, DiscreteSpace
from nav_study.core.planner import GlobalPlanner
from nav_study.core.trace import TraceRecorder
from nav_study.core.types import Cell, PlanResult, PlanStats

from ._discrete import emit_finish, reconstruct


class BFS(GlobalPlanner[Cell, "DiscreteSpace[Cell]"]):
    @property
    def name(self) -> str:
        return "bfs"

    def required_capabilities(self) -> set[Capability]:
        return {Capability.DISCRETE_SPACE}

    def plan(
        self,
        space: DiscreteSpace[Cell],
        start: Cell,
        goal: Cell,
        recorder: TraceRecorder | None = None,
    ) -> PlanResult[Cell]:
        t0 = time.monotonic()
        parent: dict[Cell, Cell] = {}
        cost_to: dict[Cell, float] = {start: 0.0}
        discovered: set[Cell] = {start}
        frontier: deque[Cell] = deque([start])
        expanded = 0
        found = False
        while frontier:
            cell = frontier.popleft()
            expanded += 1
            if recorder is not None:
                recorder.node_expanded(cell, cost_to[cell])
            if cell == goal:
                found = True
                break
            for nb, edge_cost in space.neighbors(cell):
                if nb in discovered:
                    continue
                discovered.add(nb)
                parent[nb] = cell
                cost_to[nb] = cost_to[cell] + edge_cost
                if recorder is not None:
                    recorder.edge_added(nb, cell, edge_cost)
                frontier.append(nb)

        runtime = time.monotonic() - t0
        if not found:
            emit_finish(recorder, False, [], 0.0, expanded, runtime)
            return PlanResult(success=False, stats=PlanStats(expanded_nodes=expanded))
        path = reconstruct(parent, start, goal)
        cost = cost_to[goal]
        emit_finish(recorder, True, path, cost, expanded, runtime)
        return PlanResult(True, path, cost, PlanStats(expanded_nodes=expanded))
